cmd/contour/shutdownmanager.go

```diff
@@ -0,0 +1,150 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/projectcontour/contour/internal/metrics"
+
+	"github.com/sirupsen/logrus"
+
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+const (
+	prometheusURL = "http://localhost:9001/stats/prometheus"
+	envoyAdminURL = "http://localhost:9001/healthcheck/fail"
+)
+
+type shutdownmanagerContext struct {
+	// checkInterval defines time delay between polling Envoy for open connections
+	checkInterval time.Duration
+
+	// checkDelay defines time to wait before polling Envoy for open connections
+	checkDelay time.Duration
+
+	// minOpenConnections defines the minimum amount of connections
+	// that can be open when polling for active connections in Envoy
+	minOpenConnections int
+
+	httpServePort int
+
+	logrus.FieldLogger
+}
+
+func newShutdownManagerContext() *shutdownmanagerContext {
+	// Set defaults for parameters which are then overridden via flags, ENV, or ConfigFile
+	return &shutdownmanagerContext{
+		checkInterval:      5 * time.Second,
+		checkDelay:         60 * time.Second,
+		minOpenConnections: 0,
+		httpServePort:      8090,
+	}
+}
+
+// handles the /healthz endpoint which is used for the shutdown-manager's liveness probe
+func (s *shutdownmanagerContext) healthzHandler(w http.ResponseWriter, r *http.Request) {
+	http.StatusText(http.StatusOK)
+	if _, err := w.Write([]byte("OK")); err != nil {
+		s.Error(err)
+	}
+}
+
+// handless the /shutdown endpoint which should be called from a pod preStop hook,
+// where it will block pod shutdown until envoy is able to drain connections to below the min-open threshold
+func (s *shutdownmanagerContext) shutdownHandler(w http.ResponseWriter, r *http.Request) {
+	// Send shutdown signal to Envoy to start draining connections
+	s.Infof("failing envoy healthchecks")
+	err := shutdownEnvoy(envoyAdminURL)
+	if err != nil {
+		s.Errorf("error sending envoy healthcheck fail: %v", err)
+	}
+
+	s.Infof("waiting %s before polling for draining connections", s.checkDelay)
+	time.Sleep(s.checkDelay)
+
+	for {
+		openConnections, err := getOpenConnections(prometheusURL)
+		if err != nil {
+			s.Error(err)
+		} else {
+			if openConnections <= s.minOpenConnections {
+				s.WithField("open_connections", openConnections).
+					WithField("min_connections", s.minOpenConnections).
+					Info("min number of open connections found, shutting down...")
+				return
+			}
+			s.WithField("open_connections", openConnections).
+				WithField("min_connections", s.minOpenConnections).
+				Info("polled open connections")
+		}
+		time.Sleep(s.checkInterval)
+	}
+}
+
+// shutdownEnvoy sends a POST request to /healthcheck/fail to tell Envoy to start draining connections
+func shutdownEnvoy(url string) error {
+	resp, err := http.Post(url, "", nil)
+	if err != nil {
+		return fmt.Errorf("creating post request for url %q failed: %s", url, err)
```

Comment:
Nit: s/url/URL/, here and below.

Alternatively, dropping the "url" and just saying "request for %q failed" should be just as clear.